### ISSUE TITLE
9. Ensemble_Pipeline_GridSearch

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_python-II#43]
"""
Pipeline vs Grid Search
 1. svm model
 2. 이 모델을 pipeline을 이용하여 : model workflow(dataset 전처리 -> model -> test)
 3. 이 모델을 grid search를 이용하여 : model tuning
"""
from sklearn.datasets import load_breast_cancer # dataset
from sklearn.svm import SVC # SVM 모델 생성
from sklearn.model_selection import train_test_split # split
from sklearn.preprocessing import MinMaxScaler # scaling
from sklearn.pipeline import Pipeline # model workflow
import numpy as np

### 1. SVM model
- 1) dataset load
X, y = load_breast_cancer(return_X_y = True) # x,y 동시에 리턴 받기
X.shape # (569, 30)

-열단위 평균
X.mean(axis = 0)
'''
array([1.41272917e+01, 1.92896485e+01, 9.19690334e+01, 6.54889104e+02,
       9.63602812e-02, 1.04340984e-01, 8.87993158e-02, 4.89191459e-02,
       1.81161863e-01, 6.27976098e-02, 4.05172056e-01, 1.21685343e+00,
       2.86605923e+00, 4.03370791e+01, 7.04097891e-03, 2.54781388e-02,
       3.18937163e-02, 1.17961371e-02, 2.05422988e-02, 3.79490387e-03,
       1.62691898e+01, 2.56772232e+01, 1.07261213e+02, 8.80583128e+02,
       1.32368594e-01, 2.54265044e-01, 2.72188483e-01, 1.14606223e-01,
       2.90075571e-01, 8.39458172e-02]) # 30개
1.41272917e+01   6.54889104e+02 -> 스케일링 하는 것이 좋겠어.
'''
X.min() # 0.0
X.max() # 4254.0 # 정규화스케일링 하는 것이 좋겠다.

- 2) X변수 정규화 : 전처리
scaler = MinMaxScaler().fit(X) 
X_nor = scaler.transform(X) # 2) 정규화 
X_nor.mean(axis=0)
X_nor.min() # 0.0
X_nor.max() # 1.0000000000000002

x_train, x_test,y_train, y_test = train_test_split(X_nor, y, test_size=0.3, random_state=123)

- 3) model 생성
-객체 생성
svc = SVC(gamma ='auto') 
model = svc.fit(x_train, y_train)

- 4) model 평가
score = model.score(x_test, y_test)
score # 0.9590643274853801

### 2. Pipeline : model workflow
- 1) pipeline step : [(step1:scaler),(step2:svc),.....]
 -(객체이름, 함수이름) : 함수를 실행하는 객체이름이 만들어짐.
pipe_svc = Pipeline([('scaler',MinMaxScaler()), ('svc', SVC(gamma ='auto'))])

- 2) pipeline model
model = pipe_svc.fit(x_train, y_train)

- 3) pipeline model test
score = model.score(x_test, y_test)
score # 0.9590643274853801

### 3. Grid Search : model tuneing
-pipeline -> grid search -> model turning
from sklearn.model_selection import GridSearchCV
-for문 안쓰고 그리드 서치 할수 있는 라이브러리

- 1) params 설정
params = [0.001, 0.01, 0.1, 1.0, 10.0, 100.0, 1000.0]

-dict 형식 = {'object__C' : params_range}
params_grid = [{'svc__C' : params, 'svc__kernel' : ['linear']}, # 선형에 관련된 파라미터 설정
               {'svc__C' : params, 'svc__gamma' : params, 'svc__kernel' : ['rbf']}] # 비선형에 관련된 파리미터 설정
- 2) GridSearchCV 객체
-(파이프 객체, 파람그리드, )
gs = GridSearchCV(pipe_svc, params_grid, scoring = 'accuracy', cv=10, n_jobs =1)
-scoring = 'accuracy', cv=10 : 점수평가방식/ 10차 교차검정/ n_jobs : cpu수

model = gs.fit(x_train, y_train)
acc = model.score(x_test, y_test) # 0.9883040935672515
acc

model.best_params_
-{'svc__C': 1.0, 'svc__gamma': 1.0, 'svc__kernel': 'rbf'}

-스텝4 : 파이프아니더라도 모델에 그리드서치라이브러리 사용가능
-gs = GridSearchCV(estimator = model, param_grid = params, scoring = 'accuracy', cv=5, n_jobs =-1)
